server/src/main/java/org/elasticsearch/index/mapper/SyntheticSourceDocumentParserListener.java

```diff
@@ -0,0 +1,245 @@
+            };
+            this.documents = new Stack<>();
+
+            this.currentMapper = rootMapper;
+            this.depth = 0;
+        }
+
+        public State consume(Token token) throws IOException {
+            switch (token) {
+                case Token.StartObject startObject -> {
+                    if (currentMapper instanceof ObjectMapper om && om.isEnabled() == false) {
+                        ObjectMapper parentMapper = parents.peek().parentMapper();
+                        String fullPath = parentMapper.isRoot()
+                            ? currentMapper.leafName()
+                            : parentMapper.fullPath() + "." + currentMapper.leafName();
+
+                        prepare();
+
+                        return new Storing(this, startObject, fullPath, parents.peek().parentMapper(), documents.peek().document());
+                    }
+
+                    if (currentMapper instanceof ObjectMapper om) {
+                        parents.push(new Parent(om, depth));
+                        prepare();
+                    }
+                    depth += 1;
+                }
+                case Token.EndObject endObject -> {
+                    assert depth > 0;
+
+                    if (documents.peek().depth == depth) {
+                        documents.pop();
+                    }
+
+                    depth -= 1;
+
+                    if (parents.peek().depth() == depth) {
+                        parents.pop();
+                    }
+                }
+                case Token.StartArray startArray -> {
+                    if (currentMapper instanceof ObjectMapper om) {
+                        parents.push(new Parent(om, depth));
+                        prepare();
+                    }
+                    depth += 1;
+                }
+                case Token.EndArray endArray -> {
+                    assert depth > 0;
+                    depth -= 1;
+
+                    if (parents.peek().depth() == depth) {
+                        parents.pop();
+                    }
+                }
+                case Token.FieldName fieldName -> {
+                    ObjectMapper parentMapper = parents.peek().parentMapper();
+                    currentMapper = parentMapper.getMapper(fieldName.name());
+                }
+                default -> {
+                }
+            }
+
+            return this;
+        }
+
+        /**
+         * Resets the state to prepare for new part of the document (e.g. descend into an object).
+         */
+        private void prepare() {
+            currentMapper = null;
+        }
+
+        public State consume(Event event) {
+            switch (event) {
+                case Event.DocumentSwitch ds -> documents.push(new Document(ds.document(), depth));
+            }
+
+            return this;
+        }
+
+        record Parent(ObjectMapper parentMapper, int depth) {}
+
+        record Document(LuceneDocument document, int depth) {}
```

Comment:
Why do we need depth in both? Do we need two records?

Reply:
You can have object hierarchy inside nested object and in that case depth would be different between the two. We could maybe add `LuceneDocument document` to `Parent` but then it would sometimes be `null` (unitl document switch event arrives) and i think it's an unnecessary complication.